syntax = "proto3";
import "google/protobuf/empty.proto";
package grpc_implementation;

service ServerRegistry {
    rpc RegisterServer(ServerMessage) returns (Result);
    rpc GetServerList(google.protobuf.Empty) returns (ServerList);
}

service Server {
    rpc JoinServer(Client) returns (Result);
    rpc LeaveSever(Client) returns (Result);
    rpc PublishArticle(ArticleRequest) returns (Result);
    rpc GetArticle(ArticleRequest) returns (ArticleList);
}

message ServerList{
    repeated ServerMessage serverList = 1;
}

message ArticleList {
    repeated Article articleList = 1;
}

message ArticleRequest {
    Client client = 1;
    Article article = 2;
}

message Client {
    string id = 1;
}

message ServerMessage {
    string name = 1;
    string address = 2;
}

message Article {
    oneof type {
        string _type = 1;
    }
    string author = 2;
    string time = 3;
    string content = 4;
}

enum Status {
    SUCCESS = 0;
    FAIL = 1;
}

message Result {
    Status status = 1;
}